Task 3

DavydovDenys opened the pull request.

DavydovDenys committed:
Update Car

Add to array Car instance in initialize

DavydovDenys committed:
Update Bike

Add to array Bike instance in initialize

DavydovDenys committed:
Update Transport

Add filter_by, find_by, all methods

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -15,6 +17,43 @@ def initialize(max_weight, speed, available)
+  end
+
+  def self.filter_by(attribute, &block)
+    objects_with_given_attributes = @@park.select { |p| p.respond_to?(attribute) }
```
you can use here park.select cause you already define it

DavydovDenys replied:
done

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -15,6 +17,43 @@ def initialize(max_weight, speed, available)
+
+    result = []
+    objects_with_given_attributes.each do |object|
+      if block.call(object.send(attribute))
```
better to use here public_send - read documentation about it please

DavydovDenys replied:
done

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -15,6 +17,43 @@ def initialize(max_weight, speed, available)
+
+    result = nil
+    objects_with_given_attributes.each do |object|
+      if object.send(attribute) == value
```
better to use here public_send - read documentation about it please

DavydovDenys replied:
done

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -15,6 +17,43 @@ def initialize(max_weight, speed, available)
+      end
+    end
+    result
+  end
```
you can do something like this with  'map', to simplify expression 
objects_with_given_attributes.map {|t| t if block.call(object.public_send(attribute))}

DavydovDenys replied:
done

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -15,6 +17,43 @@ def initialize(max_weight, speed, available)
+  end
+
+  def park
+    @@park
```
couple things here

def initialize - is always first method. Does not matter thich type of methods do you have
(def self.foo end or def foo end)
if you have more then 2 self methods use 'block syntax for them'
https://stackoverflow.com/questions/2505067/class-self-idiom-in-ruby
this block goes after initialize method, and all method inside without self, this construction already provide it.

DavydovDenys replied:
done

dmgrinko1 commented on lib/transport.rb:
```diff
@@ -15,6 +17,43 @@ def initialize(max_weight, speed, available)
+    @@park.select { |p| p.is_a?(self) }
+  end
+
+  def park
```
If Im not mistaken we should do something is delivery service with new attributes
@location = LOCATION[:in_garage]
    @number_of_deliveries = 0
    @delivery_cost = 0

DavydovDenys replied:
done

DavydovDenys committed:
Add changes according to comments

Replace send to public_send
Replace @@park to park
Replace each to map in .filter_by

DavydovDenys committed:
Update spec after refactoring

DavydovDenys committed:
Ad spec for '.filter_by', '.find_by', '.all'

DavydovDenys committed:
Update after Rubocop

dmgrinko1 commented:
@DavydovDenys You can merge it now
